src/Common/src/Interop/Unix/libssl/Interop.OpenSsl.cs

```diff
@@ -0,0 +1,424 @@
+                sslContext.writeBioPtr = libssl.BIO_new(memMethodWrite);
+                if ((IntPtr.Zero == sslContext.readBioPtr) || (IntPtr.Zero == sslContext.writeBioPtr))
+                {
+                    throw CreateSslException("Failed to retun new BIO for a given method type");
+                }
+
+                if (isServer)
+                {
+                    libssl.SSL_set_accept_state(sslContext.sslPtr);
+                }
+                else
+                {
+                    libssl.SSL_set_connect_state(sslContext.sslPtr);
+                }
+
+                libssl.SSL_set_bio(sslContext.sslPtr, sslContext.readBioPtr, sslContext.writeBioPtr);
+            }
+            catch
+            {
+                Marshal.StructureToPtr(sslContext, sslContextPtr, false);
+                FreeSslContext(sslContextPtr);
+                return IntPtr.Zero;
+            }
+            Marshal.StructureToPtr(sslContext, sslContextPtr, false);
+            return sslContextPtr;
+        }
+
+        internal static bool DoSslHandshake(IntPtr sslContextPtr, IntPtr recvPtr, int recvCount, out IntPtr sendPtr, out int sendCount)
+        {
+            sendPtr = IntPtr.Zero;
+            sendCount = 0;
+            SslContext context = Marshal.PtrToStructure<SslContext>(sslContextPtr);
+            bool isServer = context.isServer;
+            if ((IntPtr.Zero != recvPtr) && (recvCount > 0))
+            {
+                BioWrite(context.readBioPtr, recvPtr, recvCount);
+            }
+
+            int retVal = libssl.SSL_do_handshake(context.sslPtr);
+            if ((retVal == 1) && !isServer)
+            {            
+                return true;
+            }
+
+            int error;
+            if (retVal != 1)
+            {
+                error = GetSslError(context.sslPtr, retVal, "SSL Handshake failed: ");
+                if ((retVal != -1) || (error != libssl.SslErrorCode.SSL_ERROR_WANT_READ))
+                {
+                    throw CreateSslException(context.sslPtr, "SSL Handshake failed: ", error);
+                }
+            }
+
+            sendCount = libssl.BIO_ctrl_pending(context.writeBioPtr);
+
+            if (sendCount > 0)
+            {
+                sendPtr = Marshal.AllocHGlobal(sendCount);
+                sendCount = BioRead(context.writeBioPtr, sendPtr, sendCount);
+                if (sendCount <= 0)
+                {
+                    error = sendCount;
+                    Marshal.FreeHGlobal(sendPtr);
+                    sendPtr = IntPtr.Zero;
+                    sendCount = 0;
+                    error = GetSslError(context.sslPtr, error, "Read Bio failed: ");                 
+                    throw CreateSslException("SSL Handshake failed", error);                   
+                }
+            }
+        
+            return ((libssl.SSL_state(context.sslPtr) == libssl.SslState.SSL_ST_OK)) ? true : false;
```

Comment:
This line can be reduced to just `return libssl.SSL_state(context.sslPtr) == libssl.SslState.SSL_ST_OK;`, no?